Add "knock" feature.

Frinksy opened the pull request:
Resolves: https://github.com/ruma/ruma/issues/479

I've based it off of `next` branch since I assume this is supposed to contain breaking changes. Although I'm not sure if it should be feature-gated behind `unstable-pre-spec` since the MSC has already been merged.

Frinksy committed:
Add "knock" feature.

jplatte commented on crates/ruma-client-api/src/r0/knock/knock_room.rs:
```diff
@@ -0,0 +1,50 @@
+        pub reason: Option<&'a str>,
+
+        /// The servers to attempt to knock on the room through. One of the servers must be
+        /// participating in the room.
```
Please put the second sentence in its own paragraph.

jplatte commented on crates/ruma-client-api/src/r0.rs:
```diff
@@ -12,6 +12,7 @@ pub mod device;
 pub mod directory;
 pub mod filter;
 pub mod keys;
+pub mod knock;
```
Yes, this should be feature-gated, since it's not part of a released version of the spec yet. Maybe we should consider having separate feature flags for things that are part of the unstable spec vs. things that are only in MSCs in the future.

Frinksy replied:
I've used `unstable-pre-spec` for now.

jplatte commented on crates/ruma-client-api/src/r0/knock/knock_room.rs:
```diff
@@ -0,0 +1,50 @@
+//! [POST /_matrix/client/r0/knock/{roomIdOrAlias}](https://github.com/Sorunome/matrix-doc/blob/soru/knock/proposals/2403-knock.md)
```
Since this is documented in the unstable spec, please link to that: https://spec.matrix.org/unstable/client-server-api/#post_matrixclientr0knockroomidoralias

jplatte commented on crates/ruma-client-api/src/r0/sync/sync_events.rs:
```diff
@@ -327,6 +332,31 @@ impl Default for JoinedRoom {
 
+/// Updates to knocked rooms.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct KnockedRoom {
```
Please make the two new structs (conditionally) non-exhaustive like almost all other structs in ruma.

jplatte commented on crates/ruma-client-api/src/r0/sync/sync_events.rs:
```diff
@@ -327,6 +332,31 @@ impl Default for JoinedRoom {
+    fn default() -> Self {
+        Self { knock_state: Default::default() }
+    }
+}
```
This impl can be derived (same below).

jplatte commented on crates/ruma-federation-api/src/knock/create_knock_event_template/v1.rs:
```diff
@@ -0,0 +1,43 @@
+        authentication: ServerSignatures,
+    }
+
+    request : {
```
Inconsistent formatting

```suggestion
    request: {
```

jplatte commented on crates/ruma-federation-api/src/knock/create_knock_event_template/v1.rs:
```diff
@@ -0,0 +1,43 @@
+
+        /// The room versions the sending has support for.
+        ///
+        /// Defaults to `&[RoomVersionId::Version1]
```
```suggestion
        /// Defaults to `&[RoomVersionId::Version1]`.
```

jplatte commented on crates/ruma-federation-api/src/knock/create_knock_event_template/v1.rs:
```diff
@@ -0,0 +1,43 @@
+        /// May differ between room versions.
+        pub event: Raw<Pdu>,
+    }
+}
```
Missing constructors for `Request` and `Response`.

Frinksy committed:
Apply style changes and comments.

Feature-gate changes behind `unstable-pre-spec`.
Add missing constructors.
Fix style issues and links.

jplatte commented on crates/ruma-client-api/src/r0/sync/sync_events.rs:
```diff
@@ -333,30 +336,25 @@ impl Default for JoinedRoom {
 
 /// Updates to knocked rooms.
-#[derive(Debug, Clone, Deserialize, Serialize)]
+#[derive(Debug, Clone, Deserialize, Serialize, Default)]
```
Please order these like elsewhere: First builtin / std derives: `Clone, Debug, Default`, then `Deserialize, Serialize`.

jplatte commented on crates/ruma-federation-api/src/knock/send_knock/v1.rs:
```diff
@@ -0,0 +1,49 @@
+
+        /// The full knock event.
+        #[ruma_api(body)]
+        pub knock_event: &'a MemberEvent
```
```suggestion
        pub knock_event: &'a MemberEvent,
```

Frinksy committed:
Fix derive traits ordering and missing comma.

jplatte reviewed (approved):
LGTM, but would like to have @iinuwa review this too.

jplatte commented:
I don't see any breaking changes, can you rebase on main and change the PR base branch too?
